### PR TITLE
Fix RLP marshal byte array and enhancements to stake manager

### PR DESCRIPTION
**Pull request by Nemanja0x**
# Description

A bug is discovered in the `fastrlp` library in the `NewBytes` method. We should use `NewCopyBytes` wherever we have a non-transient slice/array, since `NewBytes` was not resetting memory properly. In a meanwhile, `fastrlp` library was fixed by wrapping `NewCopyBytes` function within the `NewBytes` (https://github.com/umbracle/fastrlp/pull/14), meaning that these two functions are interchangeable, so this PR updates the version of `fastrlp` library as well.
The stake manager is reverted to the behavior before the workaround for the bug in `fastrlp` is applied.

# Changes include

- [X] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backward-compatible and/or changes current functionality)

# Breaking changes

Please complete this section if any breaking changes have been made, otherwise delete it

# Checklist

- [X] I have assigned this PR to myself
- [ ] I have added at least 1 reviewer
- [ ] I have added the relevant labels
- [ ] I have updated the official documentation
- [ ] I have added sufficient documentation in code

## Testing

- [X] I have tested this code with the official test suite
- [ ] I have tested this code manually

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -59,7 +59,9 @@ test-e2e-polybft:
     # We can not build with race because of a bug in boltdb dependency
 	go build -o artifacts/polygon-edge .
 	env EDGE_BINARY=${PWD}/artifacts/polygon-edge E2E_TESTS=true E2E_LOGS=true \
-	go test -v -timeout=1h30m ./e2e-polybft/e2e/...
+		go test -v -timeout 1h -run TestE2E_Bridge_ChangeVotingPower github.com/0xPolygon/polygon-edge/e2e-polybft/e2e
```
Comment (goran-ethernal) [inline Makefile]: Revert this to execute all e2e tests.

File: consensus/polybft/stake_manager.go [hunk 2]
```diff
@@ -109,45 +106,32 @@ func (s *stakeManager) PostBlock(req *PostBlockRequest) error {
 	s.logger.Debug("Gotten transfer (stake changed) events from logs on block",
 		"eventsNum", len(events), "block", req.FullBlock.Block.Number())
 
+	for _, e := range events {
```
Comment (goran-ethernal) [inline consensus/polybft/stake_manager.go]: You can log this in the already existing for loop where we update the `stakeMap`.

File: consensus/polybft/stake_manager.go [hunk 3]
```diff
@@ -256,6 +242,8 @@ func (s *stakeManager) getTransferEventsFromReceipts(receipts []*types.Receipt)
 
 			var transferEvent contractsapi.TransferEvent
 
+			s.logger.Debug("Transfer event log before parse", "Log", log.Data)
```
Comment (goran-ethernal) [inline consensus/polybft/stake_manager.go]: Remove this, since we solved the bug.

File: consensus/polybft/stake_manager.go [hunk 4]
```diff
@@ -267,6 +255,8 @@ func (s *stakeManager) getTransferEventsFromReceipts(receipts []*types.Receipt)
 				continue
 			}
 
+			s.logger.Debug("Transfer event log after parse", "Log", log.Data)
```
Comment (goran-ethernal) [inline consensus/polybft/stake_manager.go]: Remove this, since we solved the bug.

File: consensus/polybft/stake_manager.go [hunk 5]
```diff
@@ -109,45 +106,32 @@ func (s *stakeManager) PostBlock(req *PostBlockRequest) error {
...
-	if len(updatedValidatorsStake) > 0 {
-		provider, err := s.blockchain.GetStateProviderForBlock(req.FullBlock.Block.Header)
-		if err != nil {
-			return err
-		}
-
-		systemState := s.blockchain.GetSystemState(provider)
-
-		for a := range updatedValidatorsStake {
-			stake, err := systemState.GetStakeOnValidatorSet(a)
```
Comment (goran-ethernal) [inline consensus/polybft/stake_manager.go:142]: Remove the `GetStakeOnValidatorSet` function from `systemState` interface, `SystemStateImpl`, and `systemStateMock`.

**Review by goran-ethernal — CHANGES_REQUESTED** [review]
I left some comments, I think should be addressed.

File: consensus/polybft/stake_manager.go (hunk 5 above)
Comment (igorcrevar) [reply consensus/polybft/stake_manager.go:142]: Fixed https://github.com/0xPolygon/polygon-edge/pull/1461/commits/b5951c8cbcae5029aeebfae91f42a8d96762bb4d

File: consensus/polybft/stake_manager.go (hunk 4 above)
Comment (igorcrevar) [reply consensus/polybft/stake_manager.go]: Fixed https://github.com/0xPolygon/polygon-edge/pull/1461/commits/b5951c8cbcae5029aeebfae91f42a8d96762bb4d

File: consensus/polybft/stake_manager.go (hunk 3 above)
Comment (igorcrevar) [reply consensus/polybft/stake_manager.go]: Fixed https://github.com/0xPolygon/polygon-edge/pull/1461/commits/b5951c8cbcae5029aeebfae91f42a8d96762bb4d

File: consensus/polybft/stake_manager.go (hunk 2 above)
Comment (igorcrevar) [reply consensus/polybft/stake_manager.go]: Fixed https://github.com/0xPolygon/polygon-edge/pull/1461/commits/b5951c8cbcae5029aeebfae91f42a8d96762bb4d

File: Makefile (hunk 1 above)
Comment (igorcrevar) [reply Makefile]: Fixed https://github.com/0xPolygon/polygon-edge/pull/1461/commits/ba933ad4577f451f5948be55c283e79ad11960c1

**Review by goran-ethernal — APPROVED** [review]
Nice :100:

File: consensus/polybft/stake_manager.go [hunk 6]
```diff
@@ -116,38 +113,25 @@ func (s *stakeManager) PostBlock(req *PostBlockRequest) error {
...
+			stakeMap.addStake(event.To, event.Value)
 		} else if event.IsUnstake() {
-			updatedValidatorsStake[event.From] = struct{}{}
+			s.logger.Debug("Unstake transfer event", "From", event.From, "Value", event.Value)
+
+			// then this amount was burned From validator address
+			stakeMap.removeStake(event.From, event.Value)
 		} else {
+			// this should not happen, but lets log it if it does
 			s.logger.Debug("Found a transfer event that represents neither stake nor unstake")
```
Comment (Stefan-Ethernal) [inline consensus/polybft/stake_manager.go]: Increase severity?
```suggestion
			s.logger.Warn("Found a transfer event that represents neither stake nor unstake")
```

File: consensus/polybft/stake_manager.go [hunk 7]
```diff
@@ -116,38 +113,25 @@ func (s *stakeManager) PostBlock(req *PostBlockRequest) error {
 
 	stakeMap := fullValidatorSet.Validators
 
-	updatedValidatorsStake := make(map[types.Address]struct{}, 0)
+	s.logger.Debug("full validator set before", "block", fullValidatorSet.BlockNumber, "data", stakeMap)
```
Comment (Stefan-Ethernal) [inline consensus/polybft/stake_manager.go:116]: Should we specify that `stakeMap` is not ordered (either change the log message or add a comment in the code). It can come as a surprise that order is not the same, but in fact we are iterating through `map`.
```suggestion
	s.logger.Debug("full validator set before", "block", fullValidatorSet.BlockNumber, "validators (unordered)", stakeMap)
```

File: consensus/polybft/stake_manager.go [hunk 8]
```diff
@@ -159,6 +143,8 @@ func (s *stakeManager) PostBlock(req *PostBlockRequest) error {
 		data.IsActive = data.VotingPower.Cmp(bigZero) > 0
 	}
 
+	s.logger.Debug("full validator set after", "block", req.FullBlock.Block.Number(), "data", stakeMap)
```
Comment (Stefan-Ethernal) [inline consensus/polybft/stake_manager.go:146]: Same here:
https://github.com/0xPolygon/polygon-edge/pull/1461#discussion_r1184695771

File: consensus/polybft/stake_manager_test.go [hunk 9]
```diff
@@ -98,7 +91,7 @@ func TestStakeManager_PostBlock(t *testing.T) {
 					stakeManager.validatorSetContract,
 					validators.getValidator(initialSetAliases[firstValidator]).Address(),
 					types.ZeroAddress,
-					zeroStake,
+					1, // remove initial 1
```
Comment (Stefan-Ethernal) [inline consensus/polybft/stake_manager_test.go]: This comment is a bit confusing? Should it be like this?
```suggestion
					1, // stake 1 initially
```

File: consensus/polybft/stake_manager_test.go [hunk 10]
```diff
@@ -512,3 +484,35 @@ func (d *dummyStakeTxRelayer) SendTransactionLocal(txn *ethgo.Transaction) (*eth
 func (d *dummyStakeTxRelayer) Client() *jsonrpc.Client {
 	return nil
 }
+
+func Test_MarshalBug(t *testing.T) {
```
Comment (Stefan-Ethernal) [inline consensus/polybft/stake_manager_test.go]: I'd move this test to `rlp_encoding_test.go`. It doesn't seem it belong here.

**Review by Stefan-Ethernal — COMMENTED** [review]
There are a few other places that uses `arena.NewBytes` in the Edge codebase. Should we change all of these to use `NewCopyBytes` instead?

File: consensus/polybft/stake_manager_test.go (hunk 10 above)
Comment (igorcrevar) [reply consensus/polybft/stake_manager_test.go]: fixed https://github.com/0xPolygon/polygon-edge/pull/1461/commits/758c3961fb1ae40fb85d7e8a871386ef4479658c

File: consensus/polybft/stake_manager_test.go (hunk 9 above)
Comment (igorcrevar) [reply consensus/polybft/stake_manager_test.go]: fixed https://github.com/0xPolygon/polygon-edge/pull/1461/commits/758c3961fb1ae40fb85d7e8a871386ef4479658c

File: consensus/polybft/stake_manager.go (hunk 8 above)
Comment (igorcrevar) [reply consensus/polybft/stake_manager.go:146]: Fixed https://github.com/0xPolygon/polygon-edge/pull/1461/commits/758c3961fb1ae40fb85d7e8a871386ef4479658c

File: consensus/polybft/stake_manager.go (hunk 7 above)
Comment (igorcrevar) [reply consensus/polybft/stake_manager.go:116]: fixed https://github.com/0xPolygon/polygon-edge/pull/1461/commits/758c3961fb1ae40fb85d7e8a871386ef4479658c

File: consensus/polybft/stake_manager.go (hunk 6 above)
Comment (igorcrevar) [reply consensus/polybft/stake_manager.go]: Fixed https://github.com/0xPolygon/polygon-edge/pull/1461/commits/758c3961fb1ae40fb85d7e8a871386ef4479658c

File: consensus/polybft/stake_manager.go (hunk 7 above)
Comment (igorcrevar) [reply consensus/polybft/stake_manager.go:116]: printed validators are now ordered (negative and zero stake validators are removed)

File: consensus/polybft/stake_manager.go (hunk 8 above)
Comment (igorcrevar) [reply consensus/polybft/stake_manager.go:146]: printed validators are now ordered (negative and zero stake validators are removed)

File: consensus/polybft/stake_manager.go [hunk 11]
```diff
@@ -353,24 +333,29 @@ func newValidatorStakeMap(validatorSet AccountSet) validatorStakeMap {
...
-// getActiveValidators returns all validators (*ValidatorMetadata) in sorted order
-func (sc validatorStakeMap) getActiveValidators(maxValidatorSetSize int) AccountSet {
+// removeStake removes given amount from validator defined by address
+func (sc *validatorStakeMap) removeStake(address types.Address, amount *big.Int) {
+	stakeData := (*sc)[address]
+	stakeData.VotingPower.Sub(stakeData.VotingPower, amount)
+	stakeData.IsActive = stakeData.VotingPower.Cmp(bigZero) > 0
+}
+
+// getSorted returns all validators (*ValidatorMetadata) in sorted order
```
Comment (Stefan-Ethernal) [inline consensus/polybft/stake_manager.go]: Maybe reword it like this, because it depends on `maxValidatorSetSize` how many are returned.
```suggestion
// getSorted returns validators (*ValidatorMetadata) in sorted order
```

**Review by Stefan-Ethernal — APPROVED** [review]
